### DISCUSSION TITLE
Migration Question on adding On Delete Cascade

### DISCUSSION THREAD

**Discussion by can-fraction** [simolus3/drift#3340] [Q&A] [answered]
Hello,
Thanks for the great package! I'm having some understanding the migration process and would appreciate any help.

 I added `on delete cascade` to my table and want to do migration. I genereted the tests and they failed as expected.
```
 Schema does not match
workout_log_table:
 columns:
  workout_id:
   Not equal: `NULL REFERENCES workout_table (id) ON DELETE CASCADE` (expected) and `NULL REFERENCES workout_table (id)` (actual) 
```
When I looked online on how to migrate this, I saw that I have to create a new temptable and switch them however I just randomly added this code and it worked. Tests are passing
```
      await m.runMigrationSteps(
        from: math.max(16, from),
        to: to,
        // ignore: missing_required_argument
        steps: migrationSteps(
          from16To18: (m, schema) async {
            await m.alterTable(TableMigration(schema.workoutLogTable));
            await m.alterTable(TableMigration(schema.workoutItemTable));
          },
        ),
      );
```
I'm not sure how my tests passed since I didnt provide any change in the function. 
Does this magically migrate or are tests misleading me? 
What kind of changes can I make on a table and the alterTable magically migrates? 
Sorry if these are basics I read the documentation but couldnt figure out.

**Comment by can-fraction** [comment, marked as answer]
Oh okay, I got lost in the web documentation but after reading the documentation in the code and reading the code, I realized it does the switch! So I guess my migration actually works. Just quick question can I migrate any table like this since essentially it switches to a new one? Like even if I add a column would the only difference be this function being slower than m.addColumn ?

**Reply by simolus3** [reply]
Drift's `alterTable` is kind of a catch-all that will re-create the table, copy data over and then delete the original table. SQLite has builtin support for `ALTER TABLE ... ADD COLUMN` which will definitely be faster, but is functionally equivalent.

Basically, `alterTable` works well when changing constraints or when adding new columns where the default data should depend on the existing row (possible with `columnTransformer`). Changing a column type with a transformer that casts also works with `alterTable`.

Personally I would always use the most expressive migration to make them easier to understand. So if you're adding a new column, `addColumn` is IMO a better choice than `alterTable`. But `alterTable` is a powerful tool for complex migrations that can't really be expressed without it (and is pretty much necessary when changing column constraints in SQLite).
